src/Components/Server/test/Circuits/ComponentHubTest.cs

```diff
@@ -0,0 +1,209 @@
+        {
+            var (mockClientProxy, hub) = InitializeComponentHub();
+
+            await hub.OnLocationChanged("https://localhost:5000/subdir/page", false);
+
+            var errorMessage = "Circuit not initialized.";
+            mockClientProxy.Verify(m => m.SendCoreAsync("JS.Error", new[] { errorMessage }, It.IsAny<CancellationToken>()), Times.Once());
+        }
+
+        private static (Mock<IClientProxy>, ComponentHub) InitializeComponentHub()
+        {
+            var ephemeralDataProtectionProvider = new EphemeralDataProtectionProvider();
+            var circuitIdFactory = new CircuitIdFactory(ephemeralDataProtectionProvider);
+            var circuitFactory = new TestCircuitFactory(
+                new Mock<IServiceScopeFactory>().Object,
+                NullLoggerFactory.Instance,
+                circuitIdFactory,
+                Options.Create(new CircuitOptions()));
+            var circuitRegistry = new CircuitRegistry(
+                Options.Create(new CircuitOptions()),
+                NullLogger<CircuitRegistry>.Instance,
+                circuitIdFactory);
+            var serializer = new TestServerComponentDeserializer();
+            var circuitHandleRegistry = new TestCircuitHandleRegistry();
+            var hub = new ComponentHub(
+                serializer: serializer,
+                dataProtectionProvider: ephemeralDataProtectionProvider,
+                circuitFactory: circuitFactory,
+                circuitIdFactory: circuitIdFactory,
+                circuitRegistry: circuitRegistry,
+                circuitHandleRegistry: circuitHandleRegistry,
+                logger: NullLogger<ComponentHub>.Instance);
+
+            // Here we mock out elements of the Hub that are typically configured
+            // by SignalR as clients connect to the hub.
+            var mockCaller = new Mock<IHubCallerClients>();
+            var mockClientProxy = new Mock<IClientProxy>();
+            mockCaller.Setup(x => x.Caller).Returns(mockClientProxy.Object);
+            hub.Clients = mockCaller.Object;
+            var mockContext = new Mock<HubCallerContext>();
+            mockContext.Setup(x => x.ConnectionId).Returns("123");
+            hub.Context = mockContext.Object;
+
+            return (mockClientProxy, hub);
+        }
+
+        private class TestCircuitHandleRegistry : ICircuitHandleRegistry
+        {
+            private bool circuitSet = false;
+
+            public CircuitHandle GetCircuitHandle(object circuitKey)
+            {
+                return null;
+            }
+
+            public CircuitHost GetCircuit(object circuitKey)
+            {
+                if (circuitSet)
+                {
+                    var serviceScope = new Mock<IServiceScope>();
+                    var circuitHost = TestCircuitHost.Create(
+                        serviceScope: new AsyncServiceScope(serviceScope.Object));
+                    return circuitHost;
```

Comment:
I'm not as familiar with the flow of these tests as you are, but is there any value in validating that the `circuitKey` is correct? For example storing the value supplied on the `SetCircuit` call and using that to decide whether to return `null` or not from here.

No big deal.

Reply:
For this particular case, I don't believe it provides a ton of value.

The circuit key is computed on a per-connection basis (so essentially on per circuit) so in real world scenarios it's not likely to be incorrect since there's only ever one value.

That and we don't really fully exercise the per-connection logic in this scenario. If there were issues with this codepath, they'd show up as failing server-side tests.